### PR TITLE
[FSDP()][9/N] Refactor ctor (continued)

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87937 [FSDP] Simplify `_reset_lazy_init()`
* #87936 [FSDP()][22/N] Refactor `_cast_buffers()` in `_lazy_init()`
* #87935 [FSDP()][21/N] Refactor `_buffer_name_to_orig_dtype` computation
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward
* #87926 [FSDP()][13/N] Refactor unshard/reshard/grads
* #87925 [FSDP()][12/N] Easy cleanup
* #87924 [FSDP()][10/N][11/N] Introduce composable (ctor only)
* **#87923 [FSDP()][9/N] Refactor ctor (continued)**

This PR makes a second pass over the constructor. The logic has been grouped into `_init_<...>` functions based on intent (e.g. `_init_prefetching_state()` or `_init_runtime_state()`). This makes the initialization code for composable FSDP much cleaner than having to re-write the same sequences of lower-level helper calls.

This PR also moves `_ExecOrderData` into its own file `_exec_order_utils.py`.

### CODE REVIEW

File: torch/distributed/fsdp/_exec_order_utils.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
+import itertools
```
Comment (awgu) [inline torch/distributed/fsdp/_exec_order_utils.py:1]: Simple move.

File: torch/distributed/fsdp/_init_utils.py [hunk 2]
```diff
@@ -22,6 +55,219 @@
 PARAM_BROADCAST_BUCKET_SIZE = int(250 * 1024 * 1024)
 FSDP_SYNCED = "_fsdp_synced"
 
+# TODO (awgu): Refactor this later
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:58]: Simple move.

File: torch/distributed/fsdp/_init_utils.py [hunk 3]
```diff
@@ -22,6 +55,219 @@
...
+    state: _State,
+    module: nn.Module,
+    ignored_modules: Optional[Iterable[torch.nn.Module]],
+) -> _State:
+    state._ignored_modules = _get_ignored_modules(module, ignored_modules)
+    state._ignored_params, state._ignored_param_names = _get_ignored_params(
+        module,
+        state._ignored_modules,
+    )
+    # TODO: FSDP's contract for buffers is not well-defined. They are
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:92]: Added this `TODO`.

File: torch/distributed/fsdp/_init_utils.py [hunk 4]
```diff
@@ -22,6 +55,219 @@
...
+    _check_single_device_module(module, state._ignored_params)
+    device_from_device_id = _get_device_from_device_id(device_id, state.rank)
+    _materialize_module(
+        module,
+        param_init_fn,
+        state._ignored_params,
+        device_from_device_id,
+        lambda _: True,
+    )
+    # TODO: We need to skip this for functional-like to avoid moving the entire
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:123]: This will get refactored in subsequent PRs.

File: torch/distributed/fsdp/_init_utils.py [hunk 5]
```diff
@@ -22,6 +55,219 @@
...
+    # Used for guarding against mistargeted forward prefetches
+    _needs_pre_forward_unshard: Dict[_HandlesKey, bool] = {}
+    state._needs_pre_forward_unshard = _needs_pre_forward_unshard
+    # The data structures use tuples of handles to generalize over the case
+    # where a module's forward involves multiple handles.
+    return state
+
+
+def _init_state_dict_state(state: _State) -> _State:
+    # TODO: after rebase
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:214]: Will do this in follow-ups.

File: torch/distributed/fsdp/_init_utils.py [hunk 6]
```diff
@@ -22,6 +55,219 @@
...
+
+@no_type_check
+def _init_param_handle_from_params(
+    state: _State,
+    params: List[nn.Parameter],
+    root_module: nn.Module,
+):
+    if len(params) == 0:
+        return
+    # TODO: Move module to GPU if needed (for non-wrapper code path) -- we need
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:244]: Refactored in subsequent PRs.

File: torch/distributed/fsdp/_wrap_utils.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            "for FSDP, where the wrapped module has batch norm submodules. "
+            "The batch norm submodules will be wrapped as separate FSDP "
+            "instances with mixed precision disabled since some batch norm "
+            "kernels do not support low precision."
+        )
+        auto_wrap_kwargs["auto_wrap_policy"] = auto_wrap_policy
+    _recursive_wrap(**auto_wrap_kwargs, **fsdp_kwargs)
+
+
+def _get_params_per_wrapped_module(
```
Comment (awgu) [inline torch/distributed/fsdp/_wrap_utils.py:60]: Added this method one PR early (will be used in the next one). This is auto wrap without actually wrapping.

File: torch/distributed/fsdp/_init_utils.py [hunk 8]
```diff
@@ -22,6 +55,219 @@
...
+    ignored_modules: Optional[Iterable[torch.nn.Module]],
+) -> _State:
+    state._ignored_modules = _get_ignored_modules(module, ignored_modules)
+    state._ignored_params, state._ignored_param_names = _get_ignored_params(
+        module,
+        state._ignored_modules,
+    )
+    # TODO: FSDP's contract for buffers is not well-defined. They are
+    # implicitly ignored for most functionality since they are not sharded;
+    # however, FSDP still imposes some semantics on buffers (e.g. buffer mixed
```
Comment (mrshenli) [inline torch/distributed/fsdp/_init_utils.py:94]: buffer is a bit hard, as we cannot anticipate how users gonna use it. It could be masks, stats, etc. We might need users to provide some hook/func to handle it.

Comment (mrshenli) [reply torch/distributed/fsdp/_init_utils.py:94]: Let's create an issue to track, but I think this is not P0 for now.

Comment (mrshenli) [reply torch/distributed/fsdp/_init_utils.py:94]: Also, DDP has the same issue. The new contract can cover all composable APIs.

File: torch/distributed/fsdp/_init_utils.py [hunk 9]
```diff
@@ -22,6 +55,219 @@
...
+    # unsharded module onto GPU before any sharding.
+    _move_module_to_device(module, state._ignored_params, device_from_device_id)
+    state.compute_device = _get_compute_device(
+        module,
+        state._ignored_params,
+        device_from_device_id,
+        state.rank,
+    )
+    state._managed_params = list(_get_orig_params(module, state._ignored_params))
+    if sync_module_states:
```
Comment (mrshenli) [inline torch/distributed/fsdp/_init_utils.py:133]: when we don't need to sync states? loaded from previous checkpoint? Any other cases?

File: torch/distributed/fsdp/_wrap_utils.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    ``auto_wrap_kwargs`` with the arguments in ``auto_wrap_kwargs`` and
+    ``fsdp_kwargs``.
+
+    Precondition: ``auto_wrap_policy`` contains the arguments expected by
+    ``_recursive_wrap()``, where ``auto_wrap_policy`` is not ``None``.
+    ``fsdp_kwargs`` contains all FSDP arguments except ``module``.
+    """
+    auto_wrap_policy = auto_wrap_kwargs["auto_wrap_policy"]
+    root_module = auto_wrap_kwargs["module"]
+    assert auto_wrap_policy is not None
```
Comment (mrshenli) [inline torch/distributed/fsdp/_wrap_utils.py:34]: can be done in followups. Any reason only assert on `auto_wrap_policy`, but not `root_module`

File: torch/distributed/fsdp/_wrap_utils.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    params_per_wrapped_module: List[List[nn.Parameter]] = []
+    visited_params = set()
+    # Constructing `wrapped_modules` with `_recursive_wrap()` orders the
+    # modules following a post-order traversal. We record parameters in
+    # `params_per_wrapped_module` using a reverse post-ordering, which is a
+    # topological sort, so that each shared parameter is guaranteed to be
+    # grouped with its lowest common ancestor module's parameters.
+    wrapped_modules.reverse()
+    wrapped_modules_set = set(wrapped_modules)
+    for module_to_wrap in wrapped_modules:
```
Comment (mrshenli) [inline torch/distributed/fsdp/_wrap_utils.py:95]: check my understanding. 

1. this for loop run in the reverse of post-order traversal and also starts from the root
2. the inner BFS basically finds all params not wrapped by any FSDP units, which should be taken care of by the current one just popped from the queue. 

> so that each shared parameter is guaranteed to be grouped with its lowest common ancestor module's parameters.

Is "lowest common ancestor module" refers to the lowest common FSDP ancestor module? If yes, then this part I didn't follow. Suppose we have a param p used by submodule A and B. A and B are wrapped by different FSDP unit. In this case, I assume p would still appear in both A's and B's flat param? And our suggestion is that user should wrap A and B in the same FSDP unit?

**Review by mrshenli — APPROVED** [review]
stamp to unblock

File: torch/distributed/fsdp/_init_utils.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/fsdp/_init_utils.py:133]: Also, I think we do not need to sync if we assume that the model initialization is deterministically the same across ranks.

File: torch/distributed/fsdp/_wrap_utils.py (hunk 10 above)
Comment (awgu) [reply torch/distributed/fsdp/_wrap_utils.py:34]: I think this was because `auto_wrap_policy` is defaulted to `None` in the FSDP constructor, so this `assert auto_wrap_policy is not None` is reflecting the intention that this function should only be run if the user specified an auto wrap policy.

File: torch/distributed/fsdp/_wrap_utils.py (hunk 11 above)
Comment (awgu) [reply torch/distributed/fsdp/_wrap_utils.py]: Thanks for catching this! This reverse-post order traversal and BFS only handles parent-child shared parameters, not sibling shared parameters.

- Parent-child shared parameters: Submodule `A` and submodule `B` are both to be logically wrapped and share a parameter `p`, and `A` is the parent of `B`. The current implementation will make sure that `p` is assigned to `A`.
- Sibling shared parameters: Submodule `A` and submodule `B` are both to be logically wrapped and share a parameter `p`, and neither `A` is the parent of `B` nor `B` is the parent of `A`. We need to rewrite the logic to traverse the module tree and pre-compute the shared parameters. Then, for each shared parameter, we need to compute the least common ancestor submodule among the ones to be logically wrapped and manually assign the shared parameter to that LCA.

I will explore this in a follow-up.

**Comment by zhaojuanmao** [comment]
ideally we still want to run the basic FSDP benchmarks to sanity check no crash, no regression before landing the PRs as it has huge changes to the core codes
